zerver/decorator.py

```diff
@@ -988,14 +987,14 @@ def wrapped_func(request: HttpRequest, *args: object, **kwargs: object) -> HttpR
                 return func(request, *args, **kwargs)
 
             user = request.user
+            remote_server = RequestNotes.get_notes(request).remote_server
 
-            if isinstance(user, AnonymousUser):
+            if settings.ZILENCER_ENABLED and remote_server is not None:
+                rate_limit_remote_server(request, remote_server, domain="api_by_remote_server")
+            elif not user.is_authenticated:
                 rate_limit_request_by_ip(request, domain="api_by_ip")
                 return func(request, *args, **kwargs)
-            elif settings.ZILENCER_ENABLED and isinstance(user, RemoteZulipServer):
-                rate_limit_remote_server(request, user, domain="api_by_remote_server")
             else:
-                assert isinstance(user, UserProfile)
```

Comment:
Makes sense. And since we are not done with the transition yet, it's still safer to check `request.user` here. I think we can safely remove this once we have the type fixed.